### ISSUE TITLE
No match of right hand side value

### ISSUE THREAD

**Issue by JustinAiken** [gaynetdinov/ex_marshal#9]
Given this session:
```ruby
{
  "session_id"=>"a8170a95eb2f56f046da8be50d92e8a5", 
  "s3_video_region"=>"us-west-1", 
  "admin_return_to"=>"/gizmo/reports/distribution_receipts",   
  "user_return_to"=>"http://usertesting.dev/admins/auth/google_oauth2/callback?state=eec87db6b1eaf789d869c6ad7def175d6d50240060b95f24&code=4/ItZbxW-FxI0TlRQelK0N-cWcUdPl1NzQIRTrI6H3AI0", 
  "warden.user.admin.key"=>[[195], nil], 
  "warden.user.admin.session"=>{"last_request_at"=>1476311578},
  "_csrf_token"=>"K+aJ46j1ZJxCgPVcxMoWQTVS+0OOKeH4NCOsvKpE+Hg=", 
  "foo"=>"bar"
}
```

I get this charlist:

```elixir
<<4, 8, 123, 13, 73, 34, 15, 115, 101, 115, 115, 105, 111, 110, 95, 105, 100, 6,
  58, 6, 69, 84, 73, 34, 37, 97, 56, 49, 55, 48, 97, 57, 53, 101, 98, 50, 102,
  53, 54, 102, 48, 52, 54, 100, 97, 56, 98, 101, 53, 48, 100, 57, 50, 101, 56,
  97, 53, 6, 59, 0, 84, 73, 34, 20, 115, 51, 95, 118, 105, 100, 101, 111, 95,
  114, 101, 103, 105, 111, 110, 6, 59, 0, 84, 73, 34, 14, 117, 115, 45, 119,
  101, 115, 116, 45, 49, 6, 59, 0, 84, 73, 34, 20, 97, 100, 109, 105, 110, 95,
  114, 101, 116, 117, 114, 110, 95, 116, 111, 6, 59, 0, 84, 73, 34, 41, 47, 103,
  105, 122, 109, 111, 47, 114, 101, 112, 111, 114, 116, 115, 47, 100, 105, 115,
  116, 114, 105, 98, 117, 116, 105, 111, 110, 95, 114, 101, 99, 101, 105, 112,
  116, 115, 6, 59, 0, 70, 73, 34, 19, 117, 115, 101, 114, 95, 114, 101, 116,
  117, 114, 110, 95, 116, 111, 6, 59, 0, 70, 73, 34, 1, 163, 104, 116, 116, 112,
  58, 47, 47, 117, 115, 101, 114, 116, 101, 115, 116, 105, 110, 103, 46, 100,
  101, 118, 47, 97, 100, 109, 105, 110, 115, 47, 97, 117, 116, 104, 47, 103,
  111, 111, 103, 108, 101, 95, 111, 97, 117, 116, 104, 50, 47, 99, 97, 108, 108,
  98, 97, 99, 107, 63, 115, 116, 97, 116, 101, 61, 101, 101, 99, 56, 55, 100,
  98, 54, 98, 49, 101, 97, 102, 55, 56, 57, 100, 56, 54, 57, 99, 54, 97, 100,
  55, 100, 101, 102, 49, 55, 53, 100, 54, 100, 53, 48, 50, 52, 48, 48, 54, 48,
  98, 57, 53, 102, 50, 52, 38, 99, 111, 100, 101, 61, 52, 47, 73, 116, 90, 98,
  120, 87, 45, 70, 120, 73, 48, 84, 108, 82, 81, 101, 108, 75, 48, 78, 45, 99,
  87, 99, 85, 100, 80, 108, 49, 78, 122, 81, 73, 82, 84, 114, 73, 54, 72, 51,
  65, 73, 48, 6, 59, 0, 84, 73, 34, 26, 119, 97, 114, 100, 101, 110, 46, 117,
  115, 101, 114, 46, 97, 100, 109, 105, 110, 46, 107, 101, 121, 6, 59, 0, 84,
  91, 7, 91, 6, 105, 1, 195, 48, 73, 34, 30, 119, 97, 114, 100, 101, 110, 46,
  117, 115, 101, 114, 46, 97, 100, 109, 105, 110, 46, 115, 101, 115, 115, 105,
  111, 110, 6, 59, 0, 84, 123, 6, 73, 34, 20, 108, 97, 115, 116, 95, 114, 101,
  113, 117, 101, 115, 116, 95, 97, 116, 6, 59, 0, 84, 108, 43, 7, 76, 145, 254,
  87, 73, 34, 16, 95, 99, 115, 114, 102, 95, 116, 111, 107, 101, 110, 6, 59, 0,
  70, 73, 34, 49, 75, 43, 97, 74, 52, 54, 106, 49, 90, 74, 120, 67, 103, 80, 86,
  99, 120, 77, 111, 87, 81, 84, 86, 83, 43, 48, 79, 79, 75, 101, 72, 52, 78, 67,
  79, 115, 118, 75, 112, 69, 43, 72, 103, 61, 6, 59, 0, 70, 73, 34, 8, 102, 111,
  111, 6, 59, 0, 84, 73, 34, 8, 98, 97, 114, 6, 59, 0, 84>>
```

and this error:

```elixir
    ** (MatchError) no match of right hand side value: ""
        lib/ex_marshal/decoder.ex:57: ExMarshal.Decoder.decode_fixnum/2
        lib/ex_marshal/decoder.ex:110: ExMarshal.Decoder.decode_string/2
        lib/ex_marshal/decoder.ex:244: ExMarshal.Decoder.decode_hash/4
        lib/ex_marshal/decoder.ex:3: ExMarshal.Decoder.decode/1
```

It's crashing somewhere parsing the long google oauth2 callback url

**Comment by gaynetdinov** [comment]
Hello.

Thank you for the issue!

I believe I fixed the issue, the issue was about the way how ExMarshal calculates string length. However the session you provided above doesn't match with bitstring

<img width="1430" alt="screen shot 2016-10-16 at 22 38 50" src="https://cloud.githubusercontent.com/assets/1541059/19420817/c0eaff6c-93f3-11e6-9697-67b8b0bb09c1.png">

Can it be that you pasted a bitstring of a slightly different session? 

I've created a branch with the fix: `fix-str-length-count`. Could you please test your session using this branch?

Thank you!

**Comment by JustinAiken** [comment]
> Can it be that you pasted a bitstring of a slightly different session?

...that's quite possible

> I've created a branch with the fix: fix-str-length-count. Could you please test your session using this branch?

Confirmed fix! 😀

**Comment by gaynetdinov** [comment]
Cool, thank you!
I just released v0.0.5 and published a new hex package version.
